### PR TITLE
Replaced older System I/O with print

### LINKED ISSUE #310: Unify the usage of `print(...)` vs `sys.std{out,err}.write`
Let's research if there's any advantage to using either of them (I'd especially take a look from testing/mockability perspective) and use one of them consistently everywhere. As for now, there's a lot of mess in this area.

### PR DESCRIPTION
**Pull request by mohitsaxenaknoldus**
Fixes #310

### CODE REVIEW

**Comment by mohitsaxenaknoldus** [comment]
It is saying no issues found? What am I missing? @PawelLipski

**Comment by PawelLipski** [comment]
> It is saying no issues found? What am I missing? @pawelprazak

You mean @PawelLipski right? ;) the end of the output says:

image

Pls bump the version in \_\_init__.py (as done on e.g. #361)

File: git_machete/client.py [hunk 1]
```diff
@@ -707,7 +707,7 @@ def traverse(
 
             if branch != current_branch and (needs_slide_out or needs_parent_sync or needs_remote_sync):
                 self.__print_new_line(False)
-                sys.stdout.write(f"Checking out {bold(branch)}\n")
+                print(f"Checking out {bold(branch)}\n")
```
Comment (PawelLipski) [inline git_machete/client.py]: Hmm is the behavior exactly the same as before the change here? doesn't `print` append `\n` automatically, unlike `sys.stdout.write`? 🤔

File: git_machete/client.py [hunk 2]
```diff
@@ -768,7 +768,7 @@ def traverse(
                         # --no-commit' (which we don't ever invoke).
                         # It's still better, however, to be on the safe side.
                         if self.__git.is_merge_in_progress():
-                            sys.stdout.write("\nMerge in progress; stopping the traversal\n")
+                            print("\nMerge in progress; stopping the traversal\n")
```
Comment (PawelLipski) [inline git_machete/client.py]: (same for the other occurrences)

File: git_machete/client.py [hunk 3]
```diff
@@ -1616,17 +1615,15 @@ def set_fork_point_override(self, branch: LocalBranchShortName, to_revision: Any
         branch_sha = self.__git.get_commit_sha_by_revision(branch)
```
Comment (PawelLipski) [inline git_machete/client.py:1615]: Looks there's still a single occurrence left:

```
$ git grep -n 'sys.stdout.write'
git_machete/client.py:1011:        sys.stdout.write(out.getvalue())
```

Comment (PawelLipski) [reply git_machete/client.py:1615]: Also! `sys.stderr.write` are still abundant... are they going to be replaced (on this PR?) as well?

**Comment by mohitsaxenaknoldus** [comment]
> > It is saying no issues found? What am I missing? @pawelprazak
> 
> You mean @PawelLipski right? ;) the end of the output says:
> 
> image
> 
> Pls bump the version in __init__.py (as done on e.g. #361)

Yes! Didn't realize there are two similar names...
Also, bumped the version.

File: git_machete/client.py (hunk 3 above)
Comment (mohitsaxenaknoldus) [reply git_machete/client.py:1615]: I've changed `sys.stderr.write()` as well but the above one I couldn't do without getting an error.

File: git_machete/cli.py [hunk 4]
```diff
@@ -790,13 +790,13 @@ def strip_remote_name(remote_branch: RemoteBranchShortName) -> LocalBranchShortN
 
     except (argparse.ArgumentError, argparse.ArgumentTypeError) as e:
         print(get_short_general_usage())
-        sys.stderr.write(f"\n{e}\n")
+        print(f"\n{e}")
```
Comment (PawelLipski) [inline git_machete/cli.py]: Buuut.... shouldn't `sys.stderr` be specified as the `file = ...`? 🤔

Comment (PawelLipski) [reply git_machete/cli.py]: Coz in the current shape, the output will go to `sys.stdout`!

Comment (mohitsaxenaknoldus) [reply git_machete/cli.py]: Like this? `print("Error in atexit._run_exitfuncs:", file=sys.stderr)`

Comment (PawelLipski) [reply git_machete/cli.py]: Yes, let's give it a shot

File: git_machete/client.py [hunk 5]
```diff
@@ -1031,7 +1030,7 @@ def print_line_prefix(branch_: LocalBranchShortName, suffix: str) -> None:
                 second_part = "Consider using `git machete fork-point --override-to=<revision>|--override-to-inferred|--override-to-parent <branch>` for each affected branch" \
                               "or reattaching the affected branches under different parent branches"
 
-            sys.stderr.write("\n")
+            print("")
```
Comment (PawelLipski) [inline git_machete/client.py]: Whoops `file=sys.stderr`?

File: git_machete/tests/functional/test_machete.py [hunk 6]
```diff
@@ -50,17 +49,17 @@ def mock_run_cmd(cmd: str, *args: str, **kwargs: Any) -> int:
     exit_code: int = completed_process.returncode
 
     if exit_code != 0:
-        sys.stderr.write(dim(f"<exit code: {exit_code}>\n\n"))
+        print(dim(f"<exit code: {exit_code}>\n"))
```
Comment (PawelLipski) [inline git_machete/tests/functional/test_machete.py]: `sys.stderr`?

File: git_machete/tests/functional/test_machete.py [hunk 7]
```diff
@@ -50,17 +49,17 @@ def mock_run_cmd(cmd: str, *args: str, **kwargs: Any) -> int:
...
 
 def mock_run_cmd_and_forward_stdout(cmd: str, *args: str, **kwargs: Any) -> int:
     completed_process: subprocess.CompletedProcess[bytes] = subprocess.run(
         [cmd] + list(args), stdout=subprocess.PIPE, stderr=subprocess.PIPE, ** kwargs)
-    sys.stdout.write(completed_process.stdout.decode('utf-8'))
+    print(completed_process.stdout.decode('utf-8'))
     exit_code: int = completed_process.returncode
     if exit_code != 0:
-        sys.stderr.write(dim(f"<exit code: {exit_code}>\n\n"))
+        print(dim(f"<exit code: {exit_code}>\n"))
```
Comment (PawelLipski) [inline git_machete/tests/functional/test_machete.py]: `sys.stderr`?

Comment (PawelLipski) [reply git_machete/tests/functional/test_machete.py]: Here and in other places!

**Comment by PawelLipski** [comment]
Also! pls retarget this PR to `develop`, as per https://github.com/VirtusLab/git-machete/blob/master/CONTRIBUTING.md#faqs-about-pull-requests

File: git_machete/client.py [hunk 8]
```diff
@@ -459,7 +459,7 @@ def get_root_of(branch: LocalBranchShortName) -> LocalBranchShortName:
             # after removing the outer layer of childless merged branches.
             # This is rare enough, however, that we can pretty much ignore this corner case.
 
-        print(bold("Discovered tree of branch dependencies:\n"))
+        print(bold("Discovered tree of branch dependencies:"))
```
Comment (PawelLipski) [inline git_machete/client.py]: Whoops this changes the output (an empty line will be missing in the output as an effect)

File: git_machete/client.py [hunk 9]
```diff
@@ -707,7 +707,7 @@ def traverse(
 
             if branch != current_branch and (needs_slide_out or needs_parent_sync or needs_remote_sync):
                 self.__print_new_line(False)
-                sys.stdout.write(f"Checking out {bold(branch)}\n")
+                print(f"Checking out {bold(branch)}", file=sys.stderr)
```
Comment (PawelLipski) [inline git_machete/client.py]: `stdout`!

File: git_machete/client.py [hunk 10]
```diff
@@ -768,7 +768,7 @@ def traverse(
                         # --no-commit' (which we don't ever invoke).
                         # It's still better, however, to be on the safe side.
                         if self.__git.is_merge_in_progress():
-                            sys.stdout.write("\nMerge in progress; stopping the traversal\n")
+                            print("\nMerge in progress; stopping the traversal", file=sys.stderr)
```
Comment (PawelLipski) [inline git_machete/client.py]: ☹️

File: git_machete/client.py [hunk 11]
```diff
@@ -848,11 +847,11 @@ def traverse(
...
             msg = f"No successor of {bold(current_branch)} needs to be slid out or synced with upstream branch or remote"
-        sys.stdout.write(f"{msg}; nothing left to update\n")
+        print(f"{msg}; nothing left to update", file=sys.stderr)
         if not any_action_suggested and initial_branch not in self.__roots:
-            sys.stdout.write(fmt("Tip: `traverse` by default starts from the current branch, "
-                                 "use flags (`--starts-from=`, `--whole` or `-w`, `-W`) to change this behavior.\n"
-                                 "Further info under `git machete traverse --help`.\n"))
+            print(fmt("Tip: `traverse` by default starts from the current branch, "
+                      "use flags (`--starts-from=`, `--whole` or `-w`, `-W`) to change this behavior.\n"
+                      "Further info under `git machete traverse --help`."), file=sys.stderr)
```
Comment (PawelLipski) [inline git_machete/client.py]: Whoops

File: git_machete/client.py [hunk 12]
```diff
@@ -1285,7 +1284,7 @@ def squash(self, *, current_branch: LocalBranchShortName, opt_fork_point: AnyRev
                 "start of range of commits to squash.")
         if len(commits) == 1:
             sha, short_sha, subject = commits[0]
-            print(f"Exactly one commit ({short_sha}) to squash, ignoring.\n")
+            print(f"Exactly one commit ({short_sha}) to squash, ignoring.")
```
Comment (PawelLipski) [inline git_machete/client.py]: A change in output!

File: git_machete/git_operations.py [hunk 13]
```diff
@@ -226,10 +225,9 @@ def get_default_editor(self) -> Optional[str]:
...
                     if name != "$" + git_machete_editor_var and self.get_config_attr_or_none('advice.macheteEditorSelection') != 'false':
                         sample_alternative = 'nano' if editor.startswith('vi') else 'vi'
-                        sys.stderr.write(
-                            fmt(f"Opening <b>{editor_repr}</b>.\n",
-                                f"To override this choice, use <b>{git_machete_editor_var}</b> env var, e.g. `export {git_machete_editor_var}={sample_alternative}`.\n\n",
-                                "See `git machete help edit` and `git machete edit --debug` for more details.\n\nUse `git config --global advice.macheteEditorSelection false` to suppress this message.\n"))
+                        print(fmt(f"Opening <b>{editor_repr}</b>.\n",
+                                  f"To override this choice, use <b>{git_machete_editor_var}</b> env var, e.g. `export {git_machete_editor_var}={sample_alternative}`.\n\n",
+                                  "See `git machete help edit` and `git machete edit --debug` for more details.\n\nUse `git config --global advice.macheteEditorSelection false` to suppress this message."))
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: `stderr` ‼️

File: git_machete/tests/functional/test_machete.py [hunk 14]
```diff
@@ -50,17 +49,17 @@ def mock_run_cmd(cmd: str, *args: str, **kwargs: Any) -> int:
...
-        sys.stderr.write(dim(f"<exit code: {exit_code}>\n\n"))
+        print(dim(f"<exit code: {exit_code}>\n"), file=sys.stderr)
     return completed_process.returncode
 
 
 def mock_run_cmd_and_forward_stdout(cmd: str, *args: str, **kwargs: Any) -> int:
     completed_process: subprocess.CompletedProcess[bytes] = subprocess.run(
         [cmd] + list(args), stdout=subprocess.PIPE, stderr=subprocess.PIPE, ** kwargs)
-    sys.stdout.write(completed_process.stdout.decode('utf-8'))
+    print(completed_process.stdout.decode('utf-8'), file=sys.stderr)
```
Comment (PawelLipski) [inline git_machete/tests/functional/test_machete.py]: `stdout`!

File: git_machete/client.py [hunk 15]
```diff
@@ -789,8 +789,7 @@ def traverse(
                         # a subsequent 'git rebase --continue'.
                         rebased_branch = self.__git.get_currently_rebased_branch_or_none()
                         if rebased_branch:  # 'remote_branch' should be equal to 'branch' at this point anyway
-                            sys.stdout.write(
-                                fmt(f"\nRebase of `{rebased_branch}` in progress; stopping the traversal\n"))
+                            print(fmt(f"\nRebase of `{rebased_branch}` in progress; stopping the traversal"), file=sys.stderr)
```
Comment (PawelLipski) [inline git_machete/client.py]: `sys.stdout`

File: git_machete/git_operations.py [hunk 16]
```diff
@@ -1,8 +1,8 @@
+import sys
 from typing import Callable, Dict, Generator, Iterator, List, Match, Optional, Set, Tuple
 
 import os
 import re
-import sys
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: Nitpick: pls revert the old position of the import

File: git_machete/git_operations.py [hunk 17]
```diff
@@ -727,10 +726,8 @@ def check_hook_executable(self, hook_path: str) -> bool:
             advice_ignored_hook = self.get_config_attr_or_none("advice.ignoredHook")
             if advice_ignored_hook != 'false':  # both empty and "true" is okay
                 # The [33m color must be used to keep consistent with how git colors this advice for its built-in hooks.
-                sys.stderr.write(
-                    colored(f"hint: The '{hook_path}' hook was ignored because it's not set as executable.", EscapeCodes.YELLOW) + "\n")
-                sys.stderr.write(
-                    colored("hint: You can disable this warning with `git config advice.ignoredHook false`.", EscapeCodes.YELLOW) + "\n")
+                print(colored(f"hint: The '{hook_path}' hook was ignored because it's not set as executable.", EscapeCodes.YELLOW))
+                print(colored("hint: You can disable this warning with `git config advice.ignoredHook false`.", EscapeCodes.YELLOW))
```
Comment (PawelLipski) [inline git_machete/git_operations.py]: `sys.stderr` for both

File: git_machete/tests/functional/test_machete.py [hunk 18]
```diff
@@ -50,17 +50,17 @@ def mock_run_cmd(cmd: str, *args: str, **kwargs: Any) -> int:
...
-        sys.stderr.write(dim(f"<exit code: {exit_code}>\n\n"))
+        print(dim(f"<exit code: {exit_code}>\n"), file=sys.stderr)
     return completed_process.returncode
 
 
 def mock_run_cmd_and_forward_stdout(cmd: str, *args: str, **kwargs: Any) -> int:
     completed_process: subprocess.CompletedProcess[bytes] = subprocess.run(
         [cmd] + list(args), stdout=subprocess.PIPE, stderr=subprocess.PIPE, ** kwargs)
-    sys.stdout.write(completed_process.stdout.decode('utf-8'))
+    print(completed_process.stdout.decode('utf-8'), file=sys.stdout)
```
Comment (PawelLipski) [inline git_machete/tests/functional/test_machete.py]: Nit: `file=sys.stdout` is redundant

File: git_machete/client.py [hunk 19]
```diff
@@ -707,7 +707,7 @@ def traverse(
 
             if branch != current_branch and (needs_slide_out or needs_parent_sync or needs_remote_sync):
                 self.__print_new_line(False)
-                sys.stdout.write(f"Checking out {bold(branch)}\n")
+                print(f"Checking out {bold(branch)}", file=sys.stdout)
```
Comment (PawelLipski) [inline git_machete/client.py]: `file` is redundant

File: git_machete/client.py [hunk 20]
```diff
@@ -768,7 +768,7 @@ def traverse(
                         # --no-commit' (which we don't ever invoke).
                         # It's still better, however, to be on the safe side.
                         if self.__git.is_merge_in_progress():
-                            sys.stdout.write("\nMerge in progress; stopping the traversal\n")
+                            print("\nMerge in progress; stopping the traversal", file=sys.stdout)
```
Comment (PawelLipski) [inline git_machete/client.py]: Ditto

File: git_machete/client.py [hunk 21]
```diff
@@ -789,8 +789,7 @@ def traverse(
                         # a subsequent 'git rebase --continue'.
                         rebased_branch = self.__git.get_currently_rebased_branch_or_none()
                         if rebased_branch:  # 'remote_branch' should be equal to 'branch' at this point anyway
-                            sys.stdout.write(
-                                fmt(f"\nRebase of `{rebased_branch}` in progress; stopping the traversal\n"))
+                            print(fmt(f"\nRebase of `{rebased_branch}` in progress; stopping the traversal"), file=sys.stdout)
```
Comment (PawelLipski) [inline git_machete/client.py]: Ditto

File: git_machete/client.py [hunk 22]
```diff
@@ -848,11 +847,11 @@ def traverse(
             msg: str = f"Reached branch {bold(current_branch)} which has no successor"
         else:
             msg = f"No successor of {bold(current_branch)} needs to be slid out or synced with upstream branch or remote"
-        sys.stdout.write(f"{msg}; nothing left to update\n")
+        print(f"{msg}; nothing left to update", file=sys.stdout)
```
Comment (PawelLipski) [inline git_machete/client.py]: Here and everywhere else!

**Review by PawelLipski — APPROVED** [review]
Approving since the required changes are just a matter of style (and not of correctness anymore)

**Comment by mohitsaxenaknoldus** [comment]
> Approving since the required changes are just a matter of style (and not of correctness anymore)

Cool! Sorry for messing up a bit. Actually, it was a bit confusing to know which was output and which was error after making them both `print`.

**Pull request merged**